CA server panic on shutdown

Got this after hitting `^C`:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4292675]

goroutine 100 [running]:
panic(0x49c93c0, 0xc8200100e0)
    /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/docker/swarm-v2/api.(*CertificateStatusResponse).Size(0x0, 0x0)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/api/ca.pb.go:721 +0x25
github.com/docker/swarm-v2/api.(*CertificateStatusResponse).Marshal(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/api/ca.pb.go:454 +0x46
github.com/docker/swarm-v2/vendor/github.com/golang/protobuf/proto.Marshal(0x5b94508, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/github.com/golang/protobuf/proto/encode.go:221 +0xca
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.protoCodec.Marshal(0x4a9a6a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/rpc_util.go:68 +0x89
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*protoCodec).Marshal(0x504b8f0, 0x4a9a6a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    <autogenerated>:20 +0xbe
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.encode(0x5b8ec18, 0x504b8f0, 0x4a9a6a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/rpc_util.go:246 +0xb4
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).sendResponse(0xc820084480, 0x69a1778, 0xc8207c82d0, 0xc8208180e0, 0x4a9a6a0, 0x0, 0x0, 0x0, 0xc8207be010, 0x0, ...)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:437 +0xba
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc820084480, 0x69a1778, 0xc8207c82d0, 0xc8208180e0, 0xc8202d65e0, 0x5041a18, 0xc820810f90, 0x0, 0x0)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:551 +0x13f5
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).handleStream(0xc820084480, 0x69a1778, 0xc8207c82d0, 0xc8208180e0, 0xc820810f90)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:682 +0x109d
github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc82080a020, 0xc820084480, 0x69a1778, 0xc8207c82d0, 0xc8208180e0)
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:348 +0xa0
created by github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
    /Users/aaronl/go/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:349 +0x9a
```

@diogomonica

--- a/ca/server.go
+++ b/ca/server.go
@@ -121,7 +121,7 @@ func (s *Server) CertificateStatus(ctx context.Context, request *api.Certificate
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		case <-s.ctx.Done():
-			return nil, ctx.Err()
+			return nil, s.ctx.Err()
 		}
 	}
 }
@@ -318,8 +318,10 @@ func (s *Server) Run(ctx context.Context) error {
 				s.updateCluster(ctx, v.Cluster)
 			}
 
+		case <-ctx.Done():
+			return ctx.Err()
 		case <-s.ctx.Done():
-			return nil
+			return s.ctx.Err()
 		}
 	}
 }

--- a/ca/testutils/cautils.go
+++ b/ca/testutils/cautils.go
@@ -113,11 +113,13 @@ func NewTestCA(t *testing.T, policy api.AcceptancePolicy) *TestCA {
 	caServer := ca.NewServer(s, managerConfig)
 	api.RegisterCAServer(grpcServer, caServer)
 
+	ctx := context.Background()
+
 	go func() {
 		grpcServer.Serve(l)
 	}()
 	go func() {
-		assert.NoError(t, caServer.Run(context.Background()))
+		caServer.Run(ctx)
 	}()
 
 	remotes := picker.NewRemotes(l.Addr().String())
@@ -132,7 +134,7 @@ func NewTestCA(t *testing.T, policy api.AcceptancePolicy) *TestCA {
 		Picker:      picker,
 		TempDir:     tempBaseDir,
 		Paths:       paths,
-		Context:     context.Background(),
+		Context:     ctx,
 		Clients:     clients,
 		Conns:       conns,
 		CAServer:    caServer,

